### DISCUSSION TITLE
Better parallelization for LowerTrs and UpperTrs (OpenMP backend)?

### DISCUSSION THREAD

**Discussion by learning-chip** [ginkgo-project/ginkgo#1122] [Ideas]
I was comparing the efficiency of the ISAI approximation (`gko::preconditioner::Isai`) against exact triangular solve (`gko::solver::LowerTrs`), in the context of ILU/IC preconditioner. While ISAI can be well parallelized (just SpMVs), the "baseline" `LowerTrs` doesn't seem to get parallel speed-up, making the comparison kind of unfair.

Looking into the omp kernel of lower_trs:

https://github.com/ginkgo-project/ginkgo/blob/693bc4fa083a09c1f737bf69e7ff69844f2204cd/omp/solver/lower_trs_kernels.cpp#L101-L119

Only the outer-loop get parallelized, which it is just for multiple right sides. No parallelism for a single RHS.

There are some fast, parallel implementations of sparse triangular solve on either [CPUs](https://github.com/IntelLabs/SpMP/blob/master/test/trsv_test.cpp) or [GPUs](https://github.com/JiyaSu/CapelliniSpTRSV) that might worth adding.

**Comment by pratikvn** [comment]
Thank you for the report. Some of our OpenMP kernels are not yet optimized. The GPU kernels with CUDA and HIP backends have optimized kernels, either with the vendor backends (cuSPARSE and hipSPARSE) or with our own kernels.

We can definitely improve our OpenMP kernels, but unfortunately it is not a very high priority for us right now. If you would like to contribute, we would be glad to help you navigate any issues you may face (with Ginkgo or OpenMP).
